### PR TITLE
Properly track values for change events on multiselect fields

### PR DESCRIPTION
**Pull request by alexgaribay**
## TLDR

Currently, multiselect fields don't trigger a change event when more than one item is selected. This PR adds tracking of mutliselect values.

## Description

Multiselect fields don't use the `value` property to reflect what the current
selected options are. Multiselects instead use the `selectedOptions` property
that uses a `HTMLCollection` with the selected nodes. Since the `HTMLCollection`
doesn't have an easy way of comparing values, we can track the selected values
by converting them to an array and then stringifying the array since arrays also
don't have an easy way to compare content.

### CODE REVIEW

File: assets/js/phoenix_live_view.js [hunk 1]
```diff
@@ -511,11 +511,36 @@ export class LiveSocket {
     for(let type of ["change", "input"]){
       this.on(type, e => {
         let input = e.target
-        let key = input.type === "checkbox" ? "checked" : "value"
-        if(this.prevInput === input && this.prevValue === input[key]){ return }
+
+        let key
```
Comment (feliperenan) [inline assets/js/phoenix_live_view.js]: Hey @alexgaribay what's up 👋 

Tested here with a multiple select input and it worked like a charm 🎉 .

Just one suggestion: I think we could get the input value as soon as we know its type so we don't need this key variable anymore, something like:

```JS
let value
if (input.type === "select-multiple") {
  value = Array.from(input.selectedOptions).map(option => option.value)
} else
  value = input.value
}
```

Not sure if this `Array.from` and `map` cause some issue here though, but it's easier to read :see_no_evil:

File: assets/js/phoenix_live_view.js [hunk 2]
```diff
@@ -511,11 +511,36 @@ export class LiveSocket {
...
+        }
+        else {
+          key = "value"
+        }
+
+        let value
+        // Handle multiselect values
+        if (input[key] instanceof HTMLCollection) {
+          let values = []
+          for (let item of input[key]) {
```
Comment (snewcomer) [inline assets/js/phoenix_live_view.js]: ie11 compat issue?  Looks like there is an existing one but I thought ie11 support was important for this pkg...

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

File: assets/js/phoenix_live_view.js (hunk 1 above)
Comment (snewcomer) [reply assets/js/phoenix_live_view.js]: I like the idea.  fyi on Array.from - https://github.com/phoenixframework/phoenix_live_view/pull/182

Comment (alexgaribay) [reply assets/js/phoenix_live_view.js]: @feliperenan I condensed to the code to follow the suggestion.

**Comment by alexgaribay** [comment]
@snewcomer I changed the loop to be ie11 compatible. Could you take another look whenever you have a chance?

File: assets/js/phoenix_live_view.js (hunk 1 above)
Comment (feliperenan) [reply assets/js/phoenix_live_view.js]: Nice 👍 

Just to know, is there some reason why you prefer the old `for` loop instead of `Array.from` and map? maybe because my approach is more expensive 🤔 

Anyway, we also could move this logic to a specific function, something like this:

```JS
let value = this.getValuesFromInput(input)

getValuesFromInput(input) {
  if (input.type === "checkbox") {
    return input["checked"]
  }
  else if (input.type === "select-multiple") {
    let selectedOptions = input["selectedOptions"]
    let values = []
    for (let i = 0; i < selectedOptions.length; i++) {
      values.push(selectedOptions.item(i).value)
    }
    // Convert to a string for easier array comparison
    return JSON.stringify(values)
  }
  else {
    return input["value"]
  }
}
```

Wdyt?

File: assets/js/phoenix_live_view.js [hunk 3]
```diff
@@ -511,11 +511,28 @@ export class LiveSocket {
     for(let type of ["change", "input"]){
       this.on(type, e => {
         let input = e.target
-        let key = input.type === "checkbox" ? "checked" : "value"
-        if(this.prevInput === input && this.prevValue === input[key]){ return }
+
+        let value
+        if (input.type === "checkbox") {
+          value = input["checked"]
```
Comment (feliperenan) [inline assets/js/phoenix_live_view.js]: Have you found some issue using `input.value`? Because testing here I haven't 🤔 

If we use `input.value` it works when the input is a checkbox as well, so we don't need this check.

```JS
if (input.type === "select-multiple") {
  let selectedOptions = input.selectedOptions
  let values = []
  for (let i = 0; i < selectedOptions.length; i++) {
    values.push(selectedOptions.item(i).value)
  }
  // Convert to a string for easier array comparison
  value = JSON.stringify(values)
}
else {
  value = input.value
}
```

File: assets/js/phoenix_live_view.js (hunk 1 above)
Comment (feliperenan) [reply assets/js/phoenix_live_view.js]: Ok, maybe because it's not IE11 compatible :/

File: assets/js/phoenix_live_view.js (hunk 3 above)
Comment (feliperenan) [reply assets/js/phoenix_live_view.js]: Nevermind, here we want to get the information if whether the input is `checked` or not. If we use `input.value` for checkbox we get different values:

```JS
checkbox.checked
=> false

checkbox.value
=> "true"
```

File: assets/js/phoenix_live_view.js [hunk 4]
```diff
@@ -511,11 +511,28 @@ export class LiveSocket {
...
+        }
+        else if (input.type === "select-multiple") {
+          let selectedOptions = input["selectedOptions"]
+          let values = []
+          for (let i = 0; i < selectedOptions.length; i++) {
+            values.push(selectedOptions.item(i).value)
+          }
+          // Convert to a string for easier array comparison
+          value = JSON.stringify(values)
```
Comment (snewcomer) [inline assets/js/phoenix_live_view.js]: Instead of `JSON.stringify`, we could do some thing like this:

```js
var selectedOptions = input["selectedOptions"]
for (let i = 0; i < selectedOptions.length; i++) {
  value += selectedOptions.item(i).value
}

```

This would avoid creating an extra data structure in the system and better perf.  What do you think?

https://jsperf.com/array-stringify/1

Comment (alexgaribay) [reply assets/js/phoenix_live_view.js]: Great call out on performance of simply concatenating. I’ll update the it with your suggestion.

Comment (alexgaribay) [reply assets/js/phoenix_live_view.js]: @snewcomer Ready for another look. Slightly deviates from the suggestion keep an array-looking string.

File: assets/js/phoenix_live_view.js [hunk 5]
```diff
@@ -514,11 +514,32 @@ export class LiveSocket {
...
+
+        let value
+        if (input.type === "checkbox") {
+          value = input["checked"]
+        }
+        else if (input.type === "select-multiple") {
+          let selectedOptions = input["selectedOptions"]
+          const lastItemIndex = selectedOptions.length - 1;
+          // Convert to a string for easier array comparison
+          value = "["
```
Comment (snewcomer) [inline assets/js/phoenix_live_view.js]: Is it ok for this key checker to be a mangled, perhaps unidentifiable string?  Or does it have to be properly put together?

Comment (alexgaribay) [reply assets/js/phoenix_live_view.js]: In the least, I wanted comma separation to ensure true value comparison. The square brackets aren't necessary for the comparison but it's an easier value to inspect.

Comment (snewcomer) [reply assets/js/phoenix_live_view.js]: Sounds good!

Looks like this is ok to merge.  What do you think?

Comment (alexgaribay) [reply assets/js/phoenix_live_view.js]: Yep. @chrismccord, good to merge?

**Comment by chrismccord** [comment]
Thank you!!! I stringified the formdata value to avoid the input type check, which should accommodate all inputs.
❤️❤️❤️🐥🔥
